Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -710,39 +1037,103 @@ DiscusMultipleScatteringCorrection::scatter(const int nScatters, Kernel::PseudoR
   // of a CSGObject sample may not generate valid tracks. Start over again
   // for this event
   if (nlinks == 0) {
-    return {false, 0, 0};
+    return {false, {0.}, 0};
   }
+  std::vector<double> weights;
   const double dl = track.front().distInsideObject;
-  const auto q = (directionToDetector - prevDirection) * kinc;
-  const auto SQ = interpolateFlat(m_SQHist, q.norm());
   if (specialSingleScatterCalc)
     vmu = 0;
   const auto AT2 = exp(-dl * vmu);
   auto scatteringXSectionFull = m_sampleShape->material().totalScatterXSection();
-  weight = weight * AT2 * SQ * scatteringXSectionFull / (4 * M_PI);
-  return {true, weight, QSS};
+  // Step through required overall energy transfer (w) values and work out what
+  // w that means for the final scatter. There will be a single w value for elastic
+  // Slightly different approach to original DISCUS code. It stepped through the w values
+  // in the supplied S(Q,w) distribution and applied each one to the final scatter. If
+  // this resulted in an overall w that equalled one of the required w values it was output.
+  // That approach implicitly assumed S(Q,w)=0 where not specified and that no interpolation
+  // on w would be needed - this may be what's required but seems possible it might not always be
+  for (auto w : wValues) {
+    const double finalE = fromWaveVector(kinc) - w;
+    if (finalE > 0) {
+      const double kout = toWaveVector(finalE);
+      const auto qVector = directionToDetector * kout - prevDirection * k;
+      const double q = qVector.norm();
+      const double finalW = fromWaveVector(k) - finalE;
+      double SQ = Interpolate2D(m_SQWS, finalW, q);
+      weights.emplace_back(weight * AT2 * SQ * scatteringXSectionFull / (4 * M_PI));
+    } else {
+      weights.emplace_back(0.);
+    }
+  }
+  return {true, weights, QSS};
+}
+
+double DiscusMultipleScatteringCorrection::getKf(const double deltaE, const double kinc) {
+  double kf;
+  if (deltaE == 0.) {
+    kf = kinc; // avoid costly sqrt
+  } else {
+    // slightly concerned that rounding errors moving between k and E may mean we take the sqrt of
+    // a negative number in here. deltaE was capped using a threshold calculated using fromWaveVector so
+    // hopefully any rounding will affect fromWaveVector(kinc) in same direction
+    kf = toWaveVector(fromWaveVector(kinc) - deltaE);
+  }
+  return kf;
+}
+
+/**
+ * Get the range of q values accessible for a particular kinc and kf. Since the kinc value is known
+ * during the simulation this is similar to direct geometry kinematics
+ * @param kf The wavevector after the scatter event
+ * @param ki The wavevector before the scatter event
+ * @return a tuple containing qmin and the qrange
+ */
+std::tuple<double, double> DiscusMultipleScatteringCorrection::getKinematicRange(double kf, double ki) {
+  double qrange;
+  const double qmin = abs(kf - ki);
+  if (ki <= kf) {
+    qrange = 2 * ki;
```

Comment:
So presumably |ki| and |kf| are determined from the energy transfer. I think this is getting the |q| range given |ki| and |kf|, the minimum being when ki and kf are parallel (|ki|-|kf|  - is using inelastic convention), the maximum being when they are anti-parallel (|ki| + |kf|)? The range being the difference of the two. But I had a couple of questions:

1.  Why is the abs value of qmin taken (surely it doesn't have to be positive)?
2. How is this effected by the crystallography convention (for example in the convention use above qrange = 2*kf, using the opposite convention it would be 2*ki, here your code seems to determine which is the appropriate one from the relative magnitudes of |ki| and |kf|, why is that the case?

Reply:
So yes the code generally handles each scatter by starting with ki, it samples an energy transfer to get kf and then works out q from ki and kf.
In the Discus code q is always >=0 so I've got the same here. I think this is done because in S(q) and S(q,w) for powders and liquids they are really S(|q|) and S(|q|,w) with only the w axis ranging across negative and positive values.
I did have a look around the crystallography convention stuff in Mantid based on a chat we had in one of our standups a couple of weeks ago. It seems like the q sign is mainly used in code that determines hkl values. Perhaps if this multiple scattering correction was ever extended to work on single crystals the assumption that q>=0 might need to change but not attempting to cover that type of instrument so far.
I'll change the code to use std::min at least but let me know if this doesn't sound right.

Reply:
One other thing about the crystallography convention\flag in Mantid. While looking into it I didn't see anything which linked this setting to the energy transfer calculation. So a positive energy transfer in Mantid is always an energy loss regardless of the crystallography convention. If there was ever any inelastic single crystal logic in Mantid this might cause a problem - not sure if that exists as a type of experiment though.

Reply:
> So yes the code generally handles each scatter by starting with ki, it samples an energy transfer to get kf and then works out q from ki and kf. In the Discus code q is always >=0 so I've got the same here. I think this is done because in S(q) and S(q,w) for powders and liquids they are really S(|q|) and S(|q|,w) with only the w axis ranging across negative and positive values. I did have a look around the crystallography convention stuff in Mantid based on a chat we had in one of our standups a couple of weeks ago. It seems like the q sign is mainly used in code that determines hkl values. Perhaps if this multiple scattering correction was ever extended to work on single crystals the assumption that q>=0 might need to change but not attempting to cover that type of instrument so far. I'll change the code to use std::min at least but let me know if this doesn't sound right.

OK sorry the convention in mantid settings is not relevant here,  it's whatever you decide in the code/workspace that you give, but it won't matter because presumably you only supply S(|q|,w) toi sample anyway (and safe to assume S(q,w) = S(-q,w)).